File: Astrolabe.Validation/Rule.cs
using System.Numerics;
using Astrolabe.Annotation;
using Astrolabe.Evaluator;

namespace Astrolabe.Validation;

[JsonString]
public enum RuleType
{
    Single,
    Multi,
    ForEach
}

[JsonBaseType("type", typeof(SingleRule))]
[JsonSubType("Rule", typeof(SingleRule))]
[JsonSubType("Rules", typeof(MultiRule))]
public abstract record Rule(RuleType Type);

public record SingleRule(EvalExpr Path, EvalExpr Props, EvalExpr Must) : Rule(RuleType.Single)
{
    public SingleRule WithProp(EvalExpr key, EvalExpr value)
    {
        return this with { Props = new CallExpr("WithProperty", [key, value, Props]) };
    }

    public SingleRule AndMust(EvalExpr andMust)
    {
        return this with { Must = Must.AndExpr(andMust) };
    }

    public SingleRule When(EvalExpr whenExpr)
    {
        return this with
        {
            Must = CallExpr.Inbuilt(InbuiltFunction.IfElse, [whenExpr, Must, ValueExpr.Null,])
        };
    }

    public SingleRule WithMessage(EvalExpr message)
    {
        return this with { Must = new CallExpr("WithMessage", [message, Must]) };
    }
}

public record MultiRule(IEnumerable<Rule> Rules) : Rule(RuleType.Multi)
{
    public static MultiRule For(params Rule[] rules)
    {
        return new MultiRule(rules);
    }

    public static Rule Concat(Rule rule1, Rule rule2)
    {
        return new MultiRule(rule1.GetRules().Concat(rule2.GetRules()));
    }
}

public record ForEachRule(EvalExpr Path, VarExpr Index, LetExpr? Variables, Rule Rule)
    : Rule(RuleType.ForEach)
{
    public ForEachRule AddRule(Rule rule)
    {
        return this with { Rule = MultiRule.Concat(Rule, rule) };
    }
}

public record ResolvedRule(DataPath Path, EvalExpr Must, IDictionary<string, object?> Properties)
{
    public T GetProperty<T>(string key)
    {
        return Properties.TryGetValue(key, out var v) ? (T)v! : default!;
    }
}

public static class RuleExtensions
{
    public static IEnumerable<Rule> GetRules(this Rule rule)
    {
        return rule switch
        {
            MultiRule multiRule => multiRule.Rules,
            _ => [rule]
        };
    }

    public static List<DataPath> GetDataOrder(this IEnumerable<ResolvedRule> rules)
    {
        var dataOrder = new List<DataPath>();
        var processed = new HashSet<DataPath>();

        var ruleList = rules.ToList();
        var ruleLookup = ruleList.ToLookup(x => x.Path);
        ruleLookup.ToList().ForEach(x => AddRules(x.Key, x));
        return dataOrder;

        void AddRules(DataPath path, IEnumerable<ResolvedRule> pathRules)
        {
            if (!processed.Add(path))
                return;
            pathRules.ToList().ForEach(x => AddExprPaths(x.Must));
            dataOrder.Add(path);
        }

        void AddPath(DataPath path)
        {
            if (!processed.Contains(path))
            {
                dataOrder.Add(path);
            }
        }

        void AddExprPaths(EvalExpr e)
        {
            switch (e)
            {
                case PathExpr { Path: var fp }:
                    AddRules(fp, ruleLookup[fp]);
                    break;
                case CallExpr { Args: var args }:
                    foreach (var expr in args)
                    {
                        AddExprPaths(expr);
                    }
                    break;
            }
        }
    }
}

File: astrolabe-validation/src/rule.ts
import {
  callExpr,
  EvalExpr,
  LetExpr,
  NullExpr,
  valueExpr,
  VarExpr,
} from "@astroapps/evaluator";

export type RuleType = "single" | "multi" | "forEach";

/**
 * Base interface for all rule types.
 */
export interface BaseRule {
  type: RuleType;
}

/**
 * A single validation rule for a specific path.
 */
export interface SingleRule extends BaseRule {
  type: "single";
  /** Expression pointing to the data field to validate */
  path: EvalExpr;
  /** Properties/metadata about the rule (stored as expression) */
  props: EvalExpr;
  /** Boolean expression that must be true for validation to pass */
  must: EvalExpr;
}

/**
 * Combines multiple rules together.
 */
export interface MultiRule extends BaseRule {
  type: "multi";
  rules: Rule[];
}

/**
 * Iterates over arrays and validates each element.
 */
export interface ForEachRule extends BaseRule {
  type: "forEach";
  /** Expression resolving to array */
  path: EvalExpr;
  /** Variable name representing current array index */
  index: VarExpr;
  /** Optional let-binding for additional variables */
  variables?: LetExpr;
  /** The rule to apply to each element */
  rule: Rule;
}

export type Rule = SingleRule | MultiRule | ForEachRule;

// ============================================================================
// Factory Functions
// ============================================================================

/**
 * Create a single validation rule.
 */
export function singleRule(
  path: EvalExpr,
  must: EvalExpr,
  props: EvalExpr = NullExpr,
): SingleRule {
  return { type: "single", path, props, must };
}

/**
 * Create a rule that combines multiple rules.
 */
export function multiRule(...rules: Rule[]): MultiRule {
  return { type: "multi", rules };
}

/**
 * Create a rule that iterates over an array and validates each element.
 */
export function forEachRule(
  path: EvalExpr,
  index: VarExpr,
  rule: Rule,
  variables?: LetExpr,
): ForEachRule {
  return { type: "forEach", path, index, variables, rule };
}

// ============================================================================
// Builder Functions (fluent API for SingleRule)
// ============================================================================

/**
 * Add a property to a single rule.
 */
export function withProp(
  rule: SingleRule,
  key: EvalExpr,
  value: EvalExpr,
): SingleRule {
  return {
    ...rule,
    props: callExpr("WithProperty", [key, value, rule.props]),
  };
}

/**
 * Add an additional condition (AND) to a single rule.
 */
export function andMust(rule: SingleRule, andMustExpr: EvalExpr): SingleRule {
  return {
    ...rule,
    must: callExpr("and", [rule.must, andMustExpr]),
  };
}

/**
 * Make a rule conditional - only validates when the condition is true.
 */
export function when(rule: SingleRule, whenExpr: EvalExpr): SingleRule {
  return {
    ...rule,
    must: callExpr("?", [whenExpr, rule.must, NullExpr]),
  };
}

/**
 * Attach a custom error message to a rule.
 */
export function withMessage(rule: SingleRule, message: EvalExpr): SingleRule {
  return {
    ...rule,
    must: callExpr("WithMessage", [message, rule.must]),
  };
}

// ============================================================================
// Utility Functions
// ============================================================================

/**
 * Get all rules from a rule (flattens MultiRule).
 */
export function getRules(rule: Rule): Rule[] {
  if (rule.type === "multi") {
    return rule.rules;
  }
  return [rule];
}

/**
 * Concatenate two rules into a single MultiRule.
 */
export function concatRules(rule1: Rule, rule2: Rule): MultiRule {
  return multiRule(...getRules(rule1), ...getRules(rule2));
}

/**
 * Add a rule to a ForEachRule.
 */
export function addRule(forEachRule: ForEachRule, rule: Rule): ForEachRule {
  return {
    ...forEachRule,
    rule: concatRules(forEachRule.rule, rule),
  };
}

// ============================================================================
// Helper for creating rules with string paths
// ============================================================================

/**
 * Create a single rule using a property expression path.
 */
export function ruleForPath(
  path: string,
  must: EvalExpr,
  props: EvalExpr = NullExpr,
): SingleRule {
  return singleRule({ type: "property", property: path }, must, props);
}

/**
 * Create a single rule using a variable expression path.
 */
export function ruleForVar(
  varName: string,
  must: EvalExpr,
  props: EvalExpr = NullExpr,
): SingleRule {
  return singleRule({ type: "var", variable: varName }, must, props);
}
